12-java-tcp-programming/COURSE_MATERIAL.md

```diff
@@ -265,258 +260,134 @@ for (int i = 0; i < dataLength; i++) {
-> be run with Java 21 or later. It is not mandatory to run this example but you
-> must understand how it works.
->
-> This example is not compatible with Java 17 but is already available in the
-> code examples repository for future use.
+Using the Docker knowledge you have acquired in the
+[Docker and Docker Compose chapter](https://github.com/heig-vd-dai-course/heig-vd-dai-course/tree/main/06-docker-and-docker-compose),
+dockerize the application.
 
-#### Answer the following questions
+The steps to dockerize the application are the following:
 
-Using the official Java documentation, can you explain the differences between
-the following different implementations? When should you use one or the other
-and why?
+- Create a `Dockerfile` for the application
+- Publish the application to GitHub Container Registry
 
-- `TcpServerSimpleTextualExample`
-- `TcpServerSingleThreadTextualExample`
-- `TcpServerMultiThreadTextualExample`
-- `TcpServerCachedThreadPoolTextualExample`
-- `TcpServerFixedThreadPoolTextualExample`
-- `TcpServerVirtualThreadTextualExample`
+You should then be able to run the server and the client in Docker containers
+and access the server from the client using the following commands:
 
-Are you able to explain why the `TcpServerSingleThreadTextualExample` does not
-work as expected?
+```sh
+# Start the server
+docker run --rm -it --name the-server <docker-image-tag> server
 
-#### Share your findings
+# Start the client and access the server container
+docker run --rm -it <docker-image-tag> client --host the-server
+```
 
-Share your results in the GitHub Discussions of this organization:
-<https://github.com/orgs/heig-vd-dai-course/discussions>.
+The `--name` sets the name of the container as well as the hostname of the
+container. This allows to access the server container using its hostname from
+the client.
 
-Create a new discussion with the following information:
+You might notice that no ports are published with the host. As both container
+run on Docker, they share the same network bridge. They can thus communicate
+together without passing by the host.
 
-- **Title**: DAI 2024-2025 - Concurrent Java TCP servers - First name Last Name
-- **Category**: Show and tell
-- **Description**: Answer the questions for this section. Add links to the
-  official Java documentation to support your answers.
+### Compare your solution with the official one
 
-This will notify us that you have completed the exercise and we can check your
-work.
+Compare your solutions with the official ones stated in the
+[Solution](#solution) section.
 
-You can compare your solution with the official one stated in the
-[Solution](#solution) section, however, **we highly recommend you to try to
-complete the practical content by yourself first to learn the most**.
+If you have any questions about the solution, feel free to ask as described in
+the [Finished? Was it easy? Was it hard?](#finished-was-it-easy-was-it-hard)
+section.
```

Comment:
I think it's a good thing to no longer ask the students to create a Github discussion, but simply to create one if needed.